lib/src/license_detection/confidence.dart

```diff
@@ -0,0 +1,236 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+import 'package:pana/src/third_party/diff_match_patch/diff.dart';
+import 'package:pana/src/license_detection/license.dart';
+import 'package:pana/src/license_detection/token_matcher.dart';
+import 'package:pana/src/license_detection/tokenizer.dart';
+
+/// Computes the confidence of [knownLicense] matching with [unknownLicense] and
+/// returns an instance of [LicenseMatch] or null based on it.
+LicenseMatch? confidenceMatch(
+  LicenseWithNGrams unknownLicense,
+  LicenseWithNGrams knownLicense,
+  MatchRange matchRange,
+  double threshold,
+) {
+  final diffs = getDiffs(
+    unknownLicense.tokens,
+    knownLicense.tokens,
+    matchRange,
+  );
+
+  final range = diffRange(tokensNormalizedValue(knownLicense.tokens), diffs);
+  final distance =
+      scoreDiffs(diffs.skip(range.start).take(range.end - range.start));
+
+  // If distance is negative it implies an we have an unaccepatable diff,
+  // therefore we return null.
+  if (distance < 0) {
+    return null;
+  }
+
+  final confidence = confidencePercentage(knownLicense.tokens.length, distance);
+
+  if (confidence >= threshold) {
+    return LicenseMatch(
+      unknownLicense.tokens
+          .skip(matchRange.input.start)
+          .take(
+            matchRange.input.end - matchRange.input.start,
+          )
+          .toList(),
+      confidence,
+      knownLicense,
+      diffs,
+      Range(range.start, range.end),
+    );
+  }
+
+  // If confidence is not greater than or equal to threshold don't
+  // return any match.
+  return null;
+}
+
+/// Calculates the confidence percentange as 1 subtracted by the ratio of
+/// Levenshtein word [distance] to the number of tokens in known license.
+@visibleForTesting
+double confidencePercentage(int knownLength, int distance) {
+  if (knownLength == 0) {
+    return 1.0;
+  }
+
+  return 1.0 - (distance / knownLength);
+}
+
+@visibleForTesting
+List<Diff> getDiffs(
+  List<Token> inputTokens,
+  List<Token> knownTokens,
+  MatchRange matchRange,
+) {
+  final dmp = DiffMatchPatch();
+
+  final unknownText = tokensNormalizedValue(inputTokens
+      .skip(matchRange.input.start)
+      .take(matchRange.input.end - matchRange.input.start));
+
+  final knownText = tokensNormalizedValue(
+    knownTokens.take(matchRange.source.end),
+  );
+
+  final diffs = dmp.diffMain(unknownText, knownText);
+
+  dmp.diffCleanupSemantic(diffs);
+
+  return List.unmodifiable(diffs);
+}
+
+/// Returns the range in [diffs] that best resembles the known license text.
```

Comment:
What kind of range? Range over text characters or tokens? Is it a range in diffs?

Reply:
Range within the `diffs` list